### PR TITLE
Add Jenkinsfile for build and test

### PR DESCRIPTION
**Pull request by andrewbonney**

### CODE REVIEW

File: Jenkinsfile [hunk 1]
```diff
@@ -0,0 +1,247 @@
...
+        booleanParam(name: "FORCE_DEBUPLOAD", defaultValue: false, description: "Force Debian package upload")
+    }
+    environment {
+        http_proxy = "http://www-cache.rd.bbc.co.uk:8080"
+        https_proxy = "http://www-cache.rd.bbc.co.uk:8080"
+    }
+    stages {
+        stage("Clean Environment") {
+            steps {
+                sh 'git clean -df'
```
Comment (samdnic) [inline Jenkinsfile]: I know this also affects the example, but do we want a `-x` as well? See https://git-scm.com/docs/git-clean#git-clean--x

File: Makefile [hunk 2]
```diff
@@ -1,66 +1,107 @@
...
 	@echo "make clean   - Get rid of scratch and byte files"
-	@echo "make test    - Tests are nice"
+	@echo "make test    - Test using tox and nose2"
+	@echo "make testenv - Create a testing environment, replacing an existing one if necessary"
+	@echo "make deb     - Create deb package"
+	@echo "make rpm     - Create rpm package"
+	@echo "make rpm_spec- Create the spec file for the rpm"
+	@echo "make wheel   - Create whl package"
+	@echo "make egg     - Create egg package"
+	@echo "make pex     - Create pex package"
```
Comment (samdnic) [inline Makefile]: Do we want to build .pex packages for nmos-common?

File: Makefile [hunk 3]
```diff
@@ -1,66 +1,107 @@
...
-nosetest:
-	tox -e py27-nose-coverage -e py3-nose-coverage
+egg:
+	$(PYTHON2) $(topdir)/setup.py bdist_egg --dist-dir=$(topbuilddir)/dist
+	$(PYTHON3) $(topdir)/setup.py bdist_egg --dist-dir=$(topbuilddir)/dist
 
-nosetest-nocover:
-	tox -e py27-nose -e py3-nose
+pex:
+	tox -c $(topdir)/tox.ini -e package
```
Comment (samdnic) [inline Makefile]: tox.ini doesn't contain a `package` env

File: tox.ini [hunk 4]
```diff
@@ -9,40 +9,40 @@
...
     coverage
     mock
 install_command = pip install --process-dependency-links {opts} {packages}
 
-[testenv:py27-test-coverage]
-commands =
-    coverage run --source=./nmoscommon -m unittest discover -s tests
-    coverage annotate
-    coverage report
+[testenv:py27-nose-coverage]
```
Comment (samdnic) [inline tox.ini]: We should talk about this in Ways of Working - we probably don't need 4 permutations of running tests...

File: setup.py [hunk 5]
```diff
@@ -121,11 +123,18 @@ def find_packages(path, base=""):
     "jsonschema>=2.3.0",
     "netifaces>=0.10.6",
     "websocket-client>=0.18.0",
-    "pybonjour>=1.1.1",
     "ujson>=1.33",
     "wsaccel>=0.6.2"
 ]
 
+# Check if setuptools is recent enough to recognise python_version syntax
```
Comment (samdnic) [inline setup.py]: A comment as to why any of this is needed is probably useful here, to call out that pybonjour doesn't support Python3

File: Makefile [hunk 6]
```diff
@@ -1,66 +1,107 @@
...
 
-.PHONY: test test-nocover nosetest nosetest-nocover clean deb install source all
+.PHONY: test test2 test3 clean install source deb dsc rpm rpm_spec wheel egg pex testenv all
```
Comment (samdnic) [inline Makefile]: There is no `test2` or `test3` target. See also earlier comments about pex.

Comment (samdnic) [reply Makefile]: Oops, first attempt at that comment got eaten by the Internet going away; apparently that's not how you edit a review!

File: Jenkinsfile (hunk 1 above)
Comment (andrewbonney) [reply Jenkinsfile]: Yes, I think we do. I'll update.

File: tox.ini (hunk 4 above)
Comment (andrewbonney) [reply tox.ini]: Agreed. Whilst people may run multiple out of curiosity, it probably just complicates things in the general case, giving us more scaffolding to keep up to date.

File: setup.py (hunk 5 above)
Comment (andrewbonney) [reply setup.py]: Will do

**Comment by andrewbonney** [comment]
Thanks. Changes made for the majority. Will await WoW discussion before changing the tox.ini

File: Jenkinsfile (hunk 1 above)
Comment (samdnic) [reply Jenkinsfile]: I'm happy to make the corresponding change to the template - my bad review didn't catch it to start with!

**Comment by simonrankine** [comment]
Happy with this now. I think we should merge this in, and deal with the tox WoW stuff once we have a decision from WoW as a separate issue, as this not being done is starting to be a bit of a pain for other repos which depend on it. May even be worth opening a bug so we can keep track of it. Is that good with @samdnic?

**Comment by samdnic** [comment]
Yup, good with me

**Pull request merged**